apis/metal3.io/v1alpha1/preprovisioningimage_types.go

```diff
@@ -0,0 +1,116 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ImageFormat enumerates the allowed image formats
+// +kubebuilder:validation:Enum=iso;initrd
+type ImageFormat string
+
+const (
+	ImageFormatISO    ImageFormat = "iso"
+	ImageFormatInitRD ImageFormat = "initrd"
+)
+
+// PreprovisioningImageSpec defines the desired state of PreprovisioningImage
+type PreprovisioningImageSpec struct {
+	// networkDataName is the name of a Secret in the local namespace that
+	// contains network data to build in to the image.
+	// +optional
+	NetworkDataName string `json:"networkDataName,omitempty"`
+
+	// architecture is the processor architecture for which to build the image.
+	// +optional
+	Architecture string `json:"architecture,omitempty"`
+}
+
+type SecretStatus struct {
+	Name    string `json:"name,omitempty"`
+	Version string `json:"version,omitempty"`
+}
+
+type ImageStatusConditionType string
+
+const (
+	// Ready indicates that the Image is available and ready to be downloaded.
+	ConditionImageReady ImageStatusConditionType = "Ready"
+
+	// Error indicates that the operator was unable to build an image.
+	ConditionImageError ImageStatusConditionType = "Error"
+)
+
+// PreprovisioningImageStatus defines the observed state of PreprovisioningImage
+type PreprovisioningImageStatus struct {
+	// imageUrl is the URL from which the built image can be downloaded.
+	ImageUrl string `json:"imageUrl,omitempty"`
```

Comment:
This may not work out for initramfs images of CoreOS specifically, since it has 3 components.

Actually, a minimum ISO image will also have 2 components.. dunno what to do about it.

Reply:
Yeah, I understand this a bit better now.
We only need to supply a URL for the part that can be customised with network config, which AIUI is only the initrd. For the minimal ISO, the rootfs gets fetched over the network, so it is not involved in network setup and can have a fixed URL.